lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,399 @@
+		case <-p.done:
+		case <-time.After(time.Second * 3):
+			slog.WarnContext(context.Background(), "Forcibly terminating networking subprocess.")
+			p.killed.Store(true)
+			p.cmd.Process.Kill()
+		}
+	}
+	return nil
+}
+
+// Dial requests a network connection from the networking subprocess.
+func (p *Process) Dial(ctx context.Context, network string, addr string) (net.Conn, error) {
+	connFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationDial,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	conn, err := net.FileConn(connFD.file)
+	if err != nil {
+		connFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childConn{
+		Conn:                    conn,
+		childNetworkingResource: connFD,
+	}, nil
+}
+
+// Listen requests a local listener from the networking subprocess.
+func (p *Process) Listen(ctx context.Context, network string, addr string) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationListen,
+		Network:   network,
+		Address:   addr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD.file)
+	if err != nil {
+		listenerFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// ListenAgent requests a local ssh-agent listener from the networking subprocess.
+func (p *Process) ListenAgent(ctx context.Context) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation: NetworkingOperationListenAgent,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD.file)
+	if err != nil {
+		listenerFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// ListenX11 requests a local XServer listener from the networking subprocess.
+func (p *Process) ListenX11(ctx context.Context, req X11Request) (net.Listener, error) {
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation:  NetworkingOperationListenX11,
+		X11Request: req,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD.file)
+	if err != nil {
+		listenerFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// childNetworkingFile is a networking file corresponding to a networking
+// resource opened in the child process, with a lifetime associated with
+// the per-request connection.
+type childNetworkingFile struct {
+	conn *net.UnixConn
+	file *os.File
+}
+
+func (f *childNetworkingFile) Close() error {
+	return trace.NewAggregate(f.conn.Close(), f.file.Close())
+}
+
+// childConn is a net.Conn backed by a connection in the child process.
+type childConn struct {
```

Comment:
Wrappers have a cost, especially when dealing with OS resources that the stdlib often type asserts on (to optimize copies into splices, for example). Doubling the amount of file descriptors for a fairly active "bastion host" that does mostly networking also has a not insignificant cost.

I'm not sure if there's a great way other than having both sides just know about the correct behavior on a per-request basis, but I would keep this cleanup-on-request-connection-drop construction only for the listeners that need it (unix listeners - which are currently not even used - and the X11 and agent listeners).